designs/serialization.md

```diff
@@ -0,0 +1,296 @@
+given type with known properties called "members", additional constraints and
+metadata called "traits", and an identifier.
+
+For each shape contained in a service, a `Schema` object will be generated that
+contains almost all of its information. Traits that are known to not affect
+serialization or deserialization will be omitted from the generated `Schema` to
+save space.
+
+Schemas will form the backbone of serialization and deserialization, carrying
+information that cannot be natively included in generated data classes.
+
+The `Schema` class will be a read-only dataclass. The following shows its basic
+definition, though the concrete definition may have a slightly different
+implementation and/or additional helper methods.
+
+```python
+@dataclass(kw_only=True, frozen=True)
+class Schema:
+    id: ShapeID
+    shape_type: ShapeType
+    traits: dict[ShapeID, "Trait"] = field(default_factory=dict)
+    members: dict[str, "Schema"] = field(default_factory=dict)
+    member_target: "Schema | None" = None
+    member_index: int | None = None
+
+    @classmethod
+    def collection(
+        cls,
+        *,
+        id: ShapeID,
+        shape_type: ShapeType = ShapeType.STRUCTURE,
+        traits: list["Trait"] | None = None,
+        members: Mapping[str, "MemberSchema"] | None = None,
+    ) -> Self:
+        ...
+
+
+@dataclass(kw_only=True, frozen=True)
+class Trait:
+    id: "ShapeID"
+    value: "DocumentValue" = field(default_factory=dict)
+```
+
+Below is an example Smithy `structure` shape, followed by the `Schema` it would
+generate.
+
+```smithy
+namespace com.example
+
+structure ExampleStructure {
+    member: Integer = 0
+}
+```
+
+```python
+EXAMPLE_STRUCTURE_SCHEMA = Schema.collection(
+    id=ShapeID("com.example#ExampleStructure"),
+    members={
+        "member": {
+            "target": INTEGER,
+            "index": 0,
+            "traits": [
+                Trait(id=ShapeID("smithy.api#default"), value=0),
+            ],
+        },
+    },
+)
+```
+
+## Shape Serializers and Serializeable Shapes
+
+Serialization will function by the interaction of two interfaces:
+`ShapeSerializer`s and `SerializeableShape`s.
+
+A `ShapeSerializer` is a class that is capable of taking a `Schema` and an
+associated shape value and serializing it in some way. For example, a
+`JSONShapeSerializer` could be written in Python to convert the shape to JSON.
+
+A `SerializeableShape` is a class that has a `serialize` method that takes a
+`ShapeSerializer` and calls the relevant methods needed to serialize it. All
+generated shapes will implement the `SerializeableShape` interface, which will
+then be the method by which all serialization is performed.
+
+Using open interfaces in this way allows for great flexibility in the generated
+Python code, which will be discussed more later.
+
+In Python, these interfaces will be represented as `Protocol`s, as shown below:
```

Comment:
Heh, I got all tripped up because Protocol in Smithy world means something specific, but in Python it seems to be kind of interface construct.

Reply:
Yeah, `Protocol`s are duck-typed interfaces essentially. You can inherit from them so they can provide default implementations, but you can't instantiate them.

Reply:
I was also confused by this for a second when I didn't see a `Protocol` class defined in the doc. Maybe we should clarify `Protocol` comes from the `typing` library in Python and maybe link to one of the following resources:
* https://docs.python.org/3/library/typing.html#typing.Protocol
* https://peps.python.org/pep-0544/